File: src/developer/forensics/feedback/attachments/metrics.cc
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/developer/forensics/feedback/attachments/metrics.h"

#include "src/developer/forensics/feedback_data/constants.h"
#include "src/developer/forensics/utils/cobalt/logger.h"
#include "src/developer/forensics/utils/cobalt/metrics.h"

namespace forensics::feedback {
namespace {

const auto* const kTimedOutMetrics = new std::map<std::string, cobalt::TimedOutData>({
    {feedback_data::kAttachmentLogKernel, cobalt::TimedOutData::kKernelLog},
    {feedback_data::kAttachmentLogSystem, cobalt::TimedOutData::kSystemLog},
    {feedback_data::kAttachmentInspect, cobalt::TimedOutData::kInspect},
});

}  // namespace

AttachmentMetrics::AttachmentMetrics(cobalt::Logger* cobalt) : cobalt_(cobalt) {}

void AttachmentMetrics::LogMetrics(const Attachments& attachments) {
  std::set<cobalt::TimedOutData> to_log;
  for (const auto& [k, v] : attachments) {
    if (v.HasError() && v.Error() == Error::kTimeout && kTimedOutMetrics->contains(k)) {
      to_log.insert(kTimedOutMetrics->at(k));
    }
  }

  for (const auto metric : to_log) {
    cobalt_->LogOccurrence(metric);
  }
}

}  // namespace forensics::feedback

File: tools/check-licenses/cmd/metrics.go
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package main

import (
	"sync"
)

type CmdMetrics struct {
	mu     sync.RWMutex
	counts map[string]int      `json:"counts"`
	values map[string][]string `json:"values"`
	files  map[string][]byte   `json:"files"`
}

var Metrics *CmdMetrics

func init() {
	Metrics = &CmdMetrics{
		counts: make(map[string]int),
		values: make(map[string][]string),
		files:  make(map[string][]byte),
	}
}

func plus1(key string) {
	Metrics.mu.Lock()
	defer Metrics.mu.Unlock()
	Metrics.counts[key] = Metrics.counts[key] + 1
}

func plusVal(key string, val string) {
	Metrics.mu.Lock()
	defer Metrics.mu.Unlock()
	Metrics.counts[key] = Metrics.counts[key] + 1
	Metrics.values[key] = append(Metrics.values[key], val)
}

func plusFile(key string, content []byte) {
	Metrics.mu.Lock()
	defer Metrics.mu.Unlock()
	Metrics.files[key] = content
}

func (m *CmdMetrics) Counts() map[string]int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.counts
}

func (m *CmdMetrics) Values() map[string][]string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.values
}

func (m *CmdMetrics) Files() map[string][]byte {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.files
}
